### ISSUE TITLE
Trouble keeping a reference to a Socket

### ISSUE THREAD

**Issue by robreuss** [IBM-Swift/BlueSocket#156]
It is unlikely this issue is specifically related to BlueSocket but I thought I would reach out anyway to see if I'm missing something about how sockets live.

My code is based on the sample.  A reference to the `EchoServer` class is maintained by a class instance (`Device`) that instantiates it and calls the `run` method.  Subsequent to a connection having been made, `Device` calls a method in the `EchoServer` class that tries to make use of the socket, a reference to which is help in an EchoServer var.  But whenever I make that subsequent class, the reference to the socket is nil.

What do I need to do to make sure my `EchoServer` class maintains the reference to the socket?

Any help would be most appreciated!

**Comment by billabt** [comment]
Any chance you can show some code?  The _EchoServer_ example keeps an array of connected socket instances that can be referenced by the instance variable _socketfd_.  Maybe your _Device_ class needs to pass the method that make use of the socket this variable so that it can be looked up in the array.  Can't be sure that's the problem with seeing some more code.

**Comment by robreuss** [comment]
This is my version of _EchoServer_ called _TCPClient_.  It is the _send_ method, called from the _Device_ class, that fails because socket is nil.  You'll note that I replaced the dictionary that you use to hold multiple sockets with a single reference.  Each _Device_ class instance will have its own instance of _TCPClient_.  

`
class TCPClient {

    static let bufferSize = 4096

    var continueRunning = true
    var socket: Socket?
    let socketLockQueue = DispatchQueue(label: "net.simplyformed.socketLockQueue")
    var delegate: TCPClientDelegate?

    init() {
        
    }
    
    deinit {
        
        logDebug("TCP Client deinit")
        // Close all open sockets...
        self.socket!.close()
        self.socket = nil
        
    }
    
    func connectTo(address: String) {
        
        let queue = DispatchQueue.global(qos: .userInteractive)
        
        queue.async { [self] in
            
            do {
                do {
                    self.socket = try Socket.create(family: Manager.tcpProtocolFamily)
                }
                catch let error {
                    // TODO: Handle error
                    logError("TCP Client got error creating socket: \(error)")
                }
                
                // TODO: We need to get the port from NetService!
                // TODO: Unwrap
                logDebug("Connecting to server at \(address):\(Manager.tcpServerPort)")
                try self.socket!.connect(to: address, port: Int32(Manager.tcpServerPort))
                
                self.addNewConnection()
            }
            catch let error {
                // TODO: Handle error
                logError("TCP Client got error connecting socket: \(error)")
            }
            catch let error {
                guard let socketError = error as? Socket.Error else {
                    print("Unexpected error...")
                    return
                }
                
                if self.continueRunning {
                    print("Error reported:\n \(socketError.description)")
                }
            }
            if self.socket!.isConnected {
                logDebug("TCP Client connected to \(address)")
                DispatchQueue.main.sync {
                    self.delegate?.connectSuccess()
                }
            }
        }
    }
    
    func send(element: Element) {
        
        do {
            // TODO: Check for nil socket and socket.connected
            logDebug("Sending element message: \(element.message)")
            try self.socket!.write(from: element.message)
        }
        catch let error {
            logError("Got error when sending from server to peripheral: \(error)")
        }
    }
    
    func addNewConnection() {

        guard let socket = self.socket else {
            logDebug("Got nil socket in addNewConnection")
            return }
        
        // Get the global concurrent queue...
        let queue = DispatchQueue.global(qos: .default)
        
        // Create the run loop work item and dispatch to the default priority global queue...
        queue.async { [unowned self, socket] in
            
            var shouldKeepRunning = true
            
            var readData = Data(capacity: Manager.TCPbufferSize)
            
            do {
                // Write the welcome string...
                //try socket.write(from: "Hello, type 'QUIT' to end session\nor 'SHUTDOWN' to stop server.\n")
                
                repeat {
                    let bytesRead = try socket.read(into: &readData)
                    
                    if bytesRead > 0 {

                        logDebug("Raw incoming data: \(readData)")
                        let (element, remainingData) = Manager.processMessage(data: readData, proto: .tcp)
                        logDebug("Element received: \(String(describing: element))")
                        
                    }
                    
                    if bytesRead == 0 {
                        logDebug("Found zero bytes, exiting")
                        shouldKeepRunning = false
                        break
                    }
                    
                    readData.count = 0
                }
                while shouldKeepRunning
                
                print("Socket: \(socket.remoteHostname):\(socket.remotePort) closed...")
                socket.close()
                
                //self.socket = nil
            }
            catch let error {
                guard let socketError = error as? Socket.Error else {
                    print("Unexpected error by connection at \(socket.remoteHostname):\(socket.remotePort)...")
                    return
                }
                if self.continueRunning {
                    print("Error reported by connection at \(socket.remoteHostname):\(socket.remotePort):\n \(socketError.description)")
                }
            }
        }
    }
    
    func shutdownServer() {
        print("\nShutdown in progress...")
        continueRunning = false
        
        self.socket!.close()
        
        DispatchQueue.main.sync {
            exit(0)
        }
    }
}
`

**Comment by robreuss** [comment]
I should mention, this is a client so it is loosely adapted from _EchoServer_!

**Comment by billabt** [comment]
Can you send me the code that calls the _send_ method along with the _Device_ class?  One thing I did notice right off the bat is that you're closing the socket in the _addNewConnection()_ function.  While this won't result in the _socket_ var being set to _nil_ since you commented out that line, it will result in subsequent read/write failures...

**Comment by robreuss** [comment]
Thanks so much for looking into this.  I've made you a collaborator for the project.  This is a very rough draft so the code might be hard to follow.  The core of the problem is in TCPClient, although I moved the connection (what was AddConnection) functionality over to the TCPServer file.  I'll look into the issue you mention above about closing the socket.
